payjoin-cli/src/app.rs

```diff
@@ -50,10 +58,158 @@ impl App {
+            payjoin::bitcoin::consensus::encode::serialize_hex(&fallback_tx)
+        );
+        let psbt = ctx
+            .process_response(&mut response.into_reader())
+            .map_err(|e| anyhow!("Failed to process response {}", e))?;
+
+        self.process_pj_response(psbt)?;
+        Ok(())
+    }
+
+    #[cfg(feature = "v2")]
+    pub async fn receive_payjoin(self, amount_arg: &str) -> Result<()> {
+        use v2::Enroller;
+
+        let mut enroller = Enroller::from_relay_config(
+            &self.config.pj_endpoint,
+            &self.config.ohttp_config,
+            &self.config.ohttp_proxy,
+        );
+        let (req, ctx) = enroller.extract_req()?;
+        log::debug!("Enrolling receiver");
+        let ohttp_response = spawn_blocking(move || {
+            let http = http_agent()?;
+            http.post(req.url.as_ref()).send_bytes(&req.body).with_context(|| "HTTP request failed")
+        })
+        .await??;
+
+        let mut enrolled = enroller
+            .process_res(ohttp_response.into_reader(), ctx)
+            .map_err(|_| anyhow!("Enrollment failed"))?;
+        log::debug!("Enrolled receiver");
+
+        let pj_uri_string =
+            self.construct_payjoin_uri(amount_arg, Some(&enrolled.fallback_target()))?;
+        println!(
+            "Listening at {}. Configured to accept payjoin at BIP 21 Payjoin Uri:",
+            self.config.pj_host
+        );
+        println!("{}", pj_uri_string);
+
+        log::debug!("Awaiting proposal");
+        let res = self.long_poll_fallback(&mut enrolled).await?;
+        log::debug!("Received request");
+        let payjoin_proposal = self
+            .process_v2_proposal(res)
+            .map_err(|e| anyhow!("Failed to process proposal {}", e))?;
+        log::debug!("Posting payjoin back");
+        let (req, ohttp_ctx) = payjoin_proposal
+            .extract_v2_req()
+            .map_err(|e| anyhow!("v2 req extraction failed {}", e))?;
+        let http = http_agent()?;
+        let res = http.post(req.url.as_str()).send_bytes(&req.body)?;
+        let mut buf = Vec::new();
+        let _ = res.into_reader().read_to_end(&mut buf)?;
+        let res = payjoin_proposal.deserialize_res(buf, ohttp_ctx);
+        log::debug!("Received response {:?}", res);
+        Ok(())
+    }
+
+    #[cfg(not(feature = "v2"))]
+    pub async fn receive_payjoin(self, amount_arg: &str) -> Result<()> {
+        let pj_uri_string = self.construct_payjoin_uri(amount_arg, None)?;
+        println!(
+            "Listening at {}. Configured to accept payjoin at BIP 21 Payjoin Uri:",
+            self.config.pj_host
+        );
+        println!("{}", pj_uri_string);
+
+        self.start_http_server().await?;
+        Ok(())
+    }
+
+    #[cfg(feature = "v2")]
+    async fn long_poll_post(&self, req_ctx: payjoin::send::RequestContext<'_>) -> Result<Psbt> {
+        loop {
+            let (req, ctx) = req_ctx.extract_v2(&self.config.ohttp_proxy)?;
```

Comment:
I suppose the loops could be contained within one function that takes this line as a closure